cache.go

```diff
@@ -0,0 +1,110 @@
+package nds
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// MultiError is returned by batch operations when there are errors with
+// particular elements. Errors will be in a one-to-one correspondence with
+// the input elements; successful elements will have a nil entry.
+type MultiError []error
+
+func (m MultiError) Error() string {
+	s, n := "", 0
+	for _, e := range m {
+		if e != nil {
+			if n == 0 {
+				s = e.Error()
+			}
+			n++
+		}
+	}
+	switch n {
+	case 0:
+		return "(0 errors)"
+	case 1:
+		return s
+	case 2:
+		return s + " (and 1 other error)"
+	}
+	return fmt.Sprintf("%s (and %d other errors)", s, n-1)
+}
+
+var (
+	// ErrCacheMiss means an item was not found in cache
+	ErrCacheMiss = errors.New("nds: cache miss")
+	// ErrCASConflict means the cache item was modified between Get and CAS calls
+	ErrCASConflict = errors.New("nds: cas conflict")
+	// ErrNotStored means that an item was not stored due to a condition check failure (e.g. during an Add or CompareAndSwap call)
+	ErrNotStored = errors.New("nds: not stored")
+)
+
+// Cacher represents a cache backend that can be used by nds.
+type Cacher interface {
+	// NewContext returns the same or a derivative of the passed in context. The returned context will
+	// be used in subsequent calls to the other Cacher functions.
+	NewContext(c context.Context) (context.Context, error)
```

Comment:
I wonder if there is a way to get rid of this `NewContext` method completely?

Reply:
Yes we could, it's really only needed if you want `nds` to manage the namespaced context in the memcache implementation. We could do this within each call or ditch the namespace completely and allow the user to use a namespaced context if required instead.

Reply:
I like the idea of ditching `NewContext` and the user using a namespaced context if required instead assuming it isn't too much of a nuisance for the API user.

Reply:
Makes sense, namespacing datastore is also left to the user via the key.